File: app/src/main/java/it/cammino/risuscito/database/dao/ConsegnatiDao.java
package it.cammino.risuscito.database.dao;

import android.arch.lifecycle.LiveData;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.Query;

import java.util.List;

import it.cammino.risuscito.database.CantoConsegnato;
import it.cammino.risuscito.database.entities.Canto;
import it.cammino.risuscito.database.entities.Consegnato;

import static android.arch.persistence.room.OnConflictStrategy.REPLACE;

@SuppressWarnings("unused")
@Dao
public interface ConsegnatiDao {

  @Query("DELETE FROM consegnato")
  void truncateTable();

  @Query("SELECT A.* FROM canto A, consegnato B WHERE A.id = B.idCanto ORDER BY a.titolo ASC")
  LiveData<List<Canto>> getLiveConsegnati();

  @Query("SELECT A.* FROM canto A, consegnato B WHERE A.id = B.idCanto ORDER BY a.titolo ASC")
  List<Canto> getConsegnati();

  @Query(
      "SELECT A.*, coalesce(B.idConsegnato,0) as consegnato FROM canto A LEFT JOIN consegnato B ON A.id = B.idCanto ORDER BY A.titolo ASC")
  LiveData<List<CantoConsegnato>> getLiveChoosen();

  @Query(
      "SELECT A.*, coalesce(B.idConsegnato,0) as consegnato FROM canto A LEFT JOIN consegnato B ON A.id = B.idCanto ORDER BY A.titolo ASC")
  List<CantoConsegnato> getChoosen();

  @Query("DELETE FROM consegnato")
  void emptyConsegnati();

  @Insert(onConflict = REPLACE)
  void insertConsegnati(Consegnato consegnato);

  @Insert(onConflict = REPLACE)
  void insertConsegnati(List<Consegnato> consegnatoList);
}

File: app/src/main/java/it/cammino/risuscito/database/dao/ConsegnatiDao.kt
package it.cammino.risuscito.database.dao

import androidx.lifecycle.LiveData
import androidx.room.*
import it.cammino.risuscito.database.entities.Consegnato
import it.cammino.risuscito.database.pojo.CantoConsegnato

@Suppress("unused")
@Dao
interface ConsegnatiDao {

    @Query("SELECT A.titolo, A.pagina, A.source, A.color, A.id, B.idConsegnato as consegnato, B.txtNota, B.numPassaggio FROM canto A, consegnato B WHERE A.id = B.idCanto")
    fun liveConsegnati(): LiveData<List<CantoConsegnato>>

    @Query("SELECT A.titolo, A.pagina, A.source, A.color, A.id, B.idConsegnato as consegnato, B.txtNota, B.numPassaggio FROM canto A, consegnato B WHERE A.id = B.idCanto ORDER BY a.titolo ASC")
    fun consegnati(): List<CantoConsegnato>

    @Query("SELECT A.titolo, A.pagina, A.source, A.color, A.id, coalesce(B.idConsegnato,-1) as consegnato, coalesce(B.txtNota, '') as txtNota, B.numPassaggio FROM canto A LEFT JOIN consegnato B ON A.id = B.idCanto ORDER BY A.titolo ASC")
    fun liveChoosen(): LiveData<List<CantoConsegnato>>

    @Query("SELECT A.titolo, A.pagina, A.source, A.color, A.id, coalesce(B.idConsegnato,-1) as consegnato, coalesce(B.txtNota, '') as txtNota, B.numPassaggio FROM canto A LEFT JOIN consegnato B ON A.id = B.idCanto")
    fun choosen(): List<CantoConsegnato>

    @Query("SELECT COALESCE((SELECT numPassaggio FROM consegnato WHERE idCanto = :id), -1)")
    fun getNumPassaggio(id: Int): Int

    @Query("SELECT * FROM consegnato")
    fun all(): List<Consegnato>

    @Query("DELETE FROM consegnato")
    fun truncateTable()

    @Query("DELETE FROM consegnato")
    fun emptyConsegnati()

    @Upsert
    fun insertConsegnati(consegnato: Consegnato)

    @Upsert
    fun insertConsegnati(consegnatoList: List<Consegnato>)

    @Update
    fun updateConsegnato(consegnato: Consegnato)
}
